ytypes/node_cache.go

```diff
@@ -0,0 +1,274 @@
+package ytypes
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	json "github.com/goccy/go-json"
+	"github.com/openconfig/ygot/util"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	gpb "github.com/openconfig/gnmi/proto/gnmi"
+)
+
+// NodeCacheOpt is an option that can be usd with calls such as `SetNode` and `GetNode`.
+//
+// The `node cache` potentially provides fast-paths for config tree traversals and offers
+// noticeable performance boosts on a busy server that calls functions such as `SetNode`
+// and `GetNode` frequently.
+//
+// Passing in the reference of the `node cache` could prevent making the `ytypes` package
+// stateful. The applications that use the `ytypes` package maintain the `node cache`.
+type NodeCacheOpt struct {
+	NodeCache *NodeCache
+}
+
+// IsGetNodeOpt implements the GetNodeOpt interface.
+func (*NodeCacheOpt) IsGetNodeOpt() {}
+
+// IsSetNodeOpt implements the SetNodeOpt interface.
+func (*NodeCacheOpt) IsSetNodeOpt() {}
+
+// IsUnmarshalOpt marks IgnoreExtraFields as a valid UnmarshalOpt.
+func (*NodeCacheOpt) IsUnmarshalOpt() {}
+
+// IsGetOrCreateNodeOpt implements the GetOrCreateNodeOpt interface.
+func (*NodeCacheOpt) IsGetOrCreateNodeOpt() {}
```

Comment:
Is this missing IsDelNodeOpt?

Reply:
I see, thanks for catching this, I think the `node.DeleteNode` should be fixed too.

Reply:
The initial cache implementation was a singleton in the `ygot` library. When it was refactored I think some bugs were introduced too. It was used by a PoC earlier but I plan to do some major work soon to improve the tests for the node cache to help us catch such issues easier.